hudi-client/src/main/java/org/apache/hudi/metadata/HoodieMetadataImpl.java

```diff
@@ -0,0 +1,1104 @@
+    ValidationUtils.checkState(!readOnly, "Metadata table cannot be committed in readonly mode");
+    //ValidationUtils.checkArgument(!recordRDD.isEmpty());
+
+    // Close all readers
+    if (basefileReader != null) {
+      basefileReader.close();
+      basefileReader = null;
+    }
+    logRecordScanner = null;
+
+    try (HoodieWriteClient writeClient = new HoodieWriteClient(jsc, config, true)) {
+      writeClient.startCommitWithTime(instantTime);
+      List<WriteStatus> statuses;
+      if (performUpdate) {
+        statuses = writeClient.upsertPreppedRecords(recordRDD, instantTime).collect();
+      } else {
+        statuses = writeClient.insertPreppedRecords(recordRDD, instantTime).collect();
+      }
+      statuses.forEach(writeStatus -> {
+        if (writeStatus.hasErrors()) {
+          throw new HoodieMetadataException("Failed to commit metadata table records at instant " + instantTime);
+        }
+      });
+    }
+  }
+
+  /**
+   * Tag each record with the location.
+   *
+   * Since we only read the latest base file in a partition, we tag the records with the instant time of the latest
+   * base file.
+   */
+  private JavaRDD<HoodieRecord> prepRecords(List<HoodieRecord> records, String partitionName) {
+    HoodieTable table = HoodieTable.create(metaClient, config, hadoopConf);
+    SliceView fsView = table.getSliceView();
+    List<HoodieBaseFile> baseFiles = fsView.getLatestFileSlices(partitionName)
+        .map(s -> s.getBaseFile())
+        .filter(b -> b.isPresent())
+        .map(b -> b.get())
+        .collect(Collectors.toList());
+
+    // All the metadata fits within a single base file
+    if (partitionName.equals(METADATA_PARTITION_NAME)) {
+      if (baseFiles.size() > 1) {
+        throw new HoodieMetadataException("Multiple base files found in metadata partition");
+      }
+    }
+
+    String fileId;
+    String instantTime;
+    if (!baseFiles.isEmpty()) {
+      fileId = baseFiles.get(0).getFileId();
+      instantTime = baseFiles.get(0).getCommitTime();
+    } else {
+      // If there is a log file then we can assume that it has the data
+      List<HoodieLogFile> logFiles = fsView.getLatestFileSlices(HoodieMetadataImpl.METADATA_PARTITION_NAME)
+          .map(s -> s.getLatestLogFile())
+          .filter(b -> b.isPresent())
+          .map(b -> b.get())
+          .collect(Collectors.toList());
+      if (logFiles.isEmpty()) {
+        // No base and log files. All are new inserts
+        return jsc.parallelize(records);
+      }
+
+      fileId = logFiles.get(0).getFileId();
+      instantTime = logFiles.get(0).getBaseCommitTime();
+    }
+
+    return jsc.parallelize(records).map(r -> r.setCurrentLocation(new HoodieRecordLocation(instantTime, fileId)));
+  }
+
+  /**
+   * Returns a list of all partitions.
+   */
+  List<String> getAllPartitionPaths() throws IOException {
+    final Timer.Context context = this.metrics.getMetadataCtx(LOOKUP_PARTITIONS_STR);
```

Comment:
metrics will be null for readonly case. It needs to be checked for null

Reply:
Initialized a HoodieMetrics object with metrics disabled for the read-only case. This should keep the code simple by not requiring if checks everywhere this.metrics is used.